### PR TITLE
fix asdict to handle Some and UNSET values

### LINKED ISSUE #4141: `strawberry.asdict` does not handle `Some`
## Describe the Bug

`strawberry.asdict` only wraps `dataclasses.asdict`, so it does not handle `Some` - dataclass inside the object is not serialized.

## Additional Context

UNSET aren't handled as well
https://github.com/strawberry-graphql/strawberry/issues/3265

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
`strawberry.asdict` was just delegating to `dataclasses.asdict`, which doesn't know about `Some` or `UNSET`. This meant `Some(value)` would end up as-is in the dict instead of being unwrapped, and `UNSET` fields would leak through instead of being excluded.

Fixed by replacing the plain `dataclasses.asdict` call with a custom recursive implementation that:
- unwraps `Some(value)` to just `value`
- skips fields that are `UNSET`
- still recursively handles nested dataclasses, lists, tuples, and dicts the same way `dataclasses.asdict` does

Fixes #4141
Related: #3265

## Summary by Sourcery

Update strawberry.asdict to correctly handle framework-specific sentinel and wrapper types while maintaining recursive conversion behavior.

Bug Fixes:
- Ensure asdict unwraps Some(value) instances to their inner values so they serialize as plain data instead of wrapper objects.
- Exclude fields set to UNSET from asdict output to prevent leaking unset sentinel values in serialized dictionaries.

Documentation:
- Document the updated asdict behavior in the public API docstring and release notes, including handling of Some and UNSET values.

Tests:
- Add tests covering Some unwrapping, UNSET field exclusion, None preservation inside Some, and nested dataclass handling for Maybe fields.

### CODE REVIEW

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

`strawberry.asdict` now correctly handles `Some` and `UNSET` values.
`Some(value)` is unwrapped to its inner value, and fields set to `UNSET`
are excluded from the resulting dictionary.



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Harikrishna KP for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```

**Comment by botberry** [comment]
Thanks for adding the `RELEASE.md` file!



Here's a preview of the changelog:

---

`strawberry.asdict` now correctly handles `Some` and `UNSET` values.
`Some(value)` is unwrapped to its inner value, and fields set to `UNSET`
are excluded from the resulting dictionary.



Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Harikrishna KP for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```

File: strawberry/types/object_type.py [hunk 1]
```diff
@@ -486,7 +490,28 @@ class User:
...
+
+    def _asdict_inner(obj: Any) -> Any:
+        if isinstance(obj, Some):
+            return _asdict_inner(obj.value)
+        if dataclasses.is_dataclass(obj) and not isinstance(obj, builtins.type):
+            result = {}
+            for f in dataclasses.fields(obj):
+                value = getattr(obj, f.name)
+                if value is UNSET:
+                    continue
```
Comment (rcybulski1122012) [inline strawberry/types/object_type.py:504]: When you use `Maybe` annotation, and value is not provided it's going to be `None`  - https://strawberry.rocks/docs/types/maybe

You need to check if the field annotation is `maybe` - AFAIR there is such a function already in strawberry. If so and the value is None, exclude it from the result.

This is the implementation I've been using in my project - it's not perfect, but may come in handy 

```python
from dataclasses import fields, is_dataclass

def strawberry_to_dict(obj) -> dict[str, Any]:
    result: dict[str, Any] = {}
    for field in fields(obj):
        value = getattr(obj, field.name)
        is_maybe = _annotation_is_maybe(field.type)

        if isinstance(value, Some):
            if is_dataclass(value.value):
                result[field.name] = _strawberry_to_dict(value.value)
            elif isinstance(value.value, list):
                result[field.name] = [_strawberry_to_dict(item) if is_dataclass(item) else item for item in value.value]
            else:
                result[field.name] = value.value
        elif value is None and not is_maybe:
            result[field.name] = None
        elif value is not UNSET and not is_maybe:
            if is_dataclass(value):
                result[field.name] = _strawberry_to_dict(value)
            elif isinstance(value, list):
                result[field.name] = [_strawberry_to_dict(item) if is_dataclass(item) else item for item in value]
            else:
                result[field.name] = value

    return result


_maybe_re = re.compile(r"^(?:strawberry\.)?Maybe\[(.+)\]$")


def _annotation_is_maybe(annotation: Any) -> bool:
    # copied from strawberry
    if isinstance(annotation, str):
        # Ideally we would try to evaluate the annotation, but the args inside
        # may still not be available, as the module is still being constructed.
        # Checking for the pattern should be good enough for now.
        return _maybe_re.match(annotation) is not None

    orig = typing.get_origin(annotation)
    if orig is typing.Annotated:
        return _annotation_is_maybe(typing.get_args(annotation)[0])
    return orig is Maybe
```

File: strawberry/types/object_type.py [hunk 2]
```diff
@@ -486,7 +490,28 @@ class User:
     # {"name": "Lorem", "age": 25}
     ```
     """
-    return dataclasses.asdict(obj)
+    from strawberry.types.maybe import Some
+    from strawberry.types.unset import UNSET
```
Comment (bellini666) [inline strawberry/types/object_type.py:494]: Are those necessary for preventing import error? If not, we can move then to module level

File: strawberry/types/object_type.py [hunk 3]
```diff
@@ -486,7 +490,28 @@ class User:
...
+
+    def _asdict_inner(obj: Any) -> Any:
+        if isinstance(obj, Some):
+            return _asdict_inner(obj.value)
+        if dataclasses.is_dataclass(obj) and not isinstance(obj, builtins.type):
```
Comment (bellini666) [inline strawberry/types/object_type.py:499]: what happens when `dataclasses.is_dataclass(obj)` and also `isinstance(obj, builtins.type)`?

**Comment by bellini666** [comment]
This got supersed by https://github.com/strawberry-graphql/strawberry/pull/4320, which solved it in a cleaner way.

Due to the inspiration in this PR, I added you as co-author of it :)
